src/device.ts

```diff
@@ -10,61 +10,102 @@
 
 'use strict';
 
-const Action = require('./action');
-const Ajv = require('ajv');
+import {Action, ActionDescription} from './action';
+import Ajv from 'ajv';
+import {Adapter} from './adapter';
+import {Property, PropertyDescription} from './property';
+import {Event, EventDescription} from './event';
 const ajv = new Ajv();
 
-class Device {
-  constructor(adapter, id) {
-    if (typeof id !== 'string') {
-      id = id.toString();
-    }
+export interface DeviceDescription {
+  id: string,
+  title: string,
+  '@context': string,
+  '@type': string[],
+  description: string,
+  properties: Record<string, PropertyDescription>,
+  actions: Record<string, ActionDescription>,
+  events: Record<string, EventDescription>,
+  links: Link[],
+  baseHref?: string,
+  pin: {
+    required: boolean,
+    pattern?: string,
+  },
+  credentialsRequired: boolean,
+}
+
+export interface Link {
+  rel: string,
+  href: string
+}
+
+export class Device {
+  private adapter: Adapter;
+
+  private id: string;
+
+  private '@context' = 'https://iot.mozilla.org/schemas';
+
+  private '@type': string[] = [];
+
+  private name: string = '';
+
+  private title = '';
+
+  private description = '';
 
+  private properties = new Map<string, Property>();
```

Comment:
Just to better understand internals. Why this is not `PropertyDescription`, similarly to `actions` and `events`?

Reply:
A property is a mutable object that can change over time.
Even if it changes it remains the same object.

An Event, on the other hand, is immutable and only exists at a certain time.
The only thing that is persistent is the description of the event.
Therefore every occurrence of an event is a new instance.

The same applies to the Action.
The `Action` class is actually an execution of the action.

TLDR;
The `Property` class holds the current state of the Property while the `Event` class represents one of many events that occur at a certain point in time.